Issue with gcc-4.9 since PTLS_ELEMENTSOF merge

A bunch of errors like this:
```
/home/travis/build/NTAP/quant/lib/deps/picotls/lib/picotls.c: In function ‘is_supported_version’:
/home/travis/build/NTAP/quant/lib/deps/picotls/lib/picotls.c:382:5: error: size of unnamed array is negative
     for (i = 0; i != PTLS_ELEMENTSOF(supported_versions); ++i)
```

--- a/include/picotls.h
+++ b/include/picotls.h
@@ -37,7 +37,8 @@ extern "C" {
 #if __GNUC__ >= 3
 #define PTLS_LIKELY(x) __builtin_expect(!!(x), 1)
 #define PTLS_UNLIKELY(x) __builtin_expect(!!(x), 0)
-#define PTLS_BUILD_ASSERT(cond) ((void)sizeof(char[2 * !!(!__builtin_constant_p(cond) || (cond)) - 1]))
+#define PTLS_BUILD_ASSERT_EXPR(cond) (sizeof(char[2 * !!(!__builtin_constant_p(cond) || (cond)) - 1]) != 0)
+#define PTLS_BUILD_ASSERT(cond) ((void)PTLS_BUILD_ASSERT_EXPR(cond))
 #else
 #define PTLS_LIKELY(x) (x)
 #define PTLS_UNLIKELY(x) (x)
@@ -46,12 +47,12 @@ extern "C" {
 
 /* __builtin_types_compatible_p yields incorrect results when older versions of GCC is used; see #303 */
 #if defined(__clang__) || __GNUC__ >= 6
-#define PTLS_ASSERT_IS_ARRAY(a) PTLS_BUILD_ASSERT(__builtin_types_compatible_p(__typeof__(a[0])[], __typeof__(a)))
+#define PTLS_ASSERT_IS_ARRAY_EXPR(a) PTLS_BUILD_ASSERT_EXPR(__builtin_types_compatible_p(__typeof__(a[0])[], __typeof__(a)))
 #else
-#define PTLS_ASSERT_IS_ARRAY(a) 1
+#define PTLS_ASSERT_IS_ARRAY_EXPR(a) 1
 #endif
 
-#define PTLS_ELEMENTSOF(x) (PTLS_ASSERT_IS_ARRAY(x), sizeof(x) / sizeof((x)[0]))
+#define PTLS_ELEMENTSOF(x) (PTLS_ASSERT_IS_ARRAY_EXPR(x) * sizeof(x) / sizeof((x)[0]))
 
 #ifdef _WINDOWS
 #define PTLS_THREADLOCAL __declspec(thread)

--- a/include/picotls.h
+++ b/include/picotls.h
@@ -38,11 +38,16 @@ extern "C" {
 #define PTLS_LIKELY(x) __builtin_expect(!!(x), 1)
 #define PTLS_UNLIKELY(x) __builtin_expect(!!(x), 0)
 #define PTLS_BUILD_ASSERT(cond) ((void)sizeof(char[2 * !!(!__builtin_constant_p(cond) || (cond)) - 1]))
-#define PTLS_ASSERT_IS_ARRAY(a) PTLS_BUILD_ASSERT(__builtin_types_compatible_p(__typeof__(a[0])[], __typeof__(a)))
 #else
 #define PTLS_LIKELY(x) (x)
 #define PTLS_UNLIKELY(x) (x)
 #define PTLS_BUILD_ASSERT(cond) 1
+#endif
+
+/* __builtin_types_compatible_p yields incorrect results when older versions of GCC is used; see #303 */
+#if defined(__clang__) || __GNUC__ >= 6
+#define PTLS_ASSERT_IS_ARRAY(a) PTLS_BUILD_ASSERT(__builtin_types_compatible_p(__typeof__(a[0])[], __typeof__(a)))
+#else
 #define PTLS_ASSERT_IS_ARRAY(a) 1
 #endif
 

--- a/t/cli.c
+++ b/t/cli.c
@@ -180,7 +180,7 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
         }
 
         /* encrypt data to send, if any is available */
-        if (encbuf.off == 0) {
+        if (encbuf.off == 0 || state == IN_HANDSHAKE) {
             static const size_t block_size = 16384;
             if (inputfd >= 0 && (FD_ISSET(inputfd, &readfds) || FD_ISSET(inputfd, &exceptfds))) {
                 if ((ret = ptls_buffer_reserve(&ptbuf, block_size)) != 0)

--- a/t/cli.c
+++ b/t/cli.c
@@ -73,7 +73,6 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
 
     ptls_t *tls = ptls_new(ctx, server_name == NULL);
     ptls_buffer_t rbuf, encbuf, ptbuf;
-    char bytebuf[16384] = {0};
     enum { IN_HANDSHAKE, IN_1RTT, IN_SHUTDOWN } state = IN_HANDSHAKE;
     int inputfd = 0, ret = 0;
     size_t early_bytes_sent = 0;
@@ -133,6 +132,7 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
 
         /* consume incoming messages */
         if (FD_ISSET(sockfd, &readfds) || FD_ISSET(sockfd, &exceptfds)) {
+            char bytebuf[16384];
             size_t off = 0, leftlen;
             while ((ioret = read(sockfd, bytebuf, sizeof(bytebuf))) == -1 && errno == EINTR)
                 ;
@@ -152,13 +152,6 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
                             shift_buffer(&ptbuf, early_bytes_sent);
                         if (request_key_update)
                             ptls_update_key(tls, 1);
-                        if (ptbuf.off != 0) {
-                            if ((ret = ptls_send(tls, &encbuf, ptbuf.base, ptbuf.off)) != 0) {
-                                fprintf(stderr, "ptls_send(1rtt):%d\n", ret);
-                                goto Exit;
-                            }
-                            ptbuf.off = 0;
-                        }
                     } else if (ret == PTLS_ERROR_IN_PROGRESS) {
                         /* ok */
                     } else {
@@ -187,12 +180,15 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
         }
 
         /* encrypt data to send, if any is available */
-        ioret = 0;
         if (encbuf.off == 0) {
+            static const size_t block_size = 16384;
             if (inputfd >= 0 && (FD_ISSET(inputfd, &readfds) || FD_ISSET(inputfd, &exceptfds))) {
-                while ((ioret = read(inputfd, bytebuf, sizeof(bytebuf))) == -1 && errno == EINTR)
+                if ((ret = ptls_buffer_reserve(&ptbuf, block_size)) != 0)
+                    goto Exit;
+                while ((ioret = read(inputfd, ptbuf.base + ptbuf.off, block_size)) == -1 && errno == EINTR)
                     ;
                 if (ioret > 0) {
+                    ptbuf.off += ioret;
                 } else if (ioret == 0) {
                     /* closed */
                     if (input_file != NULL)
@@ -200,11 +196,15 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
                     inputfd = -1;
                 }
             } else if (inputfd == inputfd_is_benchmark) {
-                ioret = sizeof(bytebuf);
+                if (ptbuf.capacity < block_size) {
+                    if ((ret = ptls_buffer_reserve(&ptbuf, block_size - ptbuf.capacity)) != 0)
+                        goto Exit;
+                    memset(ptbuf.base + ptbuf.capacity, 0, block_size - ptbuf.capacity);
+                }
+                ptbuf.off = block_size;
             }
         }
-        if (ioret > 0) {
-            ptls_buffer_pushv(&ptbuf, bytebuf, ioret);
+        if (ptbuf.off != 0) {
             if (state == IN_HANDSHAKE) {
                 size_t send_amount = 0;
                 if (server_name != NULL && hsprop->client.max_early_data_size != NULL) {
@@ -221,7 +221,7 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
                     early_bytes_sent += send_amount;
                 }
             } else {
-                if ((ret = ptls_send(tls, &encbuf, bytebuf, ioret)) != 0) {
+                if ((ret = ptls_send(tls, &encbuf, ptbuf.base, ptbuf.off)) != 0) {
                     fprintf(stderr, "ptls_send(1rtt):%d\n", ret);
                     goto Exit;
                 }

